### DISCUSSION TITLE
Bug in postgresql@14

### DISCUSSION THREAD

**Discussion by dhendry** [Homebrew/discussions#3579] [Tap maintenance and Homebrew development]
# Overview/TLDR

Im trying to debug an issue with the psycopg2 but suspect I have found a bug in the postgresql@14 package. I wanted to just post here first before filing a bug in the hopes somebody can sanity check that this looks wrong.

Essentially it looks like the postgresql@14 as of 2022-08-29 appears to have a bad lib path:

![image](https://user-images.githubusercontent.com/2521996/187225353-584bce27-539d-4d9d-b38b-98f15af2d873.png)

I would not expect to need to use `/opt/homebrew/Cellar/postgresql@14/14.5_1/lib/**postgresql@14**/...` to access the library files. Can somebody just confirm?

If this is a bug I suspect it is related to the following cc: @MikeMcQuaid:
* https://github.com/Homebrew/homebrew-core/commit/8060df9169055ee1fb9d1d69fd149c3922e864ba
* Which appears to come form https://github.com/Homebrew/homebrew-core/pull/107726 

# Additional background

* For legacy reasons we still need to use the python package psycopg2== 2.8.6.
* On M1 macs, this needs to be built from source
* After updating brew, we started seeing the following failure

![image](https://user-images.githubusercontent.com/2521996/187229180-70f5faf9-7a62-406e-ab99-ad5eaca7d01a.png)

* It appears the library cannot find `libpq.5.dylib` in the expected paths, one of which should be valid except for the additional `postgresql@14` path component which I would not expect to be necessary as it seems to be a brew internal construct

**Comment by carlocab** [comment]
Yea, the libraries should probably be installed into `lib` rather than `lib/postgresql@14`. CC @bayandin

Also:

Please [don't use screenshots](https://meta.stackoverflow.com/a/285557) when sharing the output of commands in your terminal. They're often difficult to read and impossible to reuse (e.g. when looking up error messages). Instead, copy the text here and wrap them in three back-ticks (```) on the line before and on the line after to format them properly.

**Reply by dhendry** [reply]
Thanks, 

Created bug: https://github.com/Homebrew/homebrew-core/issues/109164

**Reply by bayandin** [reply]
I believe this is expected changes we done to support multiple Postgres version with extensions. You should use `pg_config` to get a correct path to lib and include dirs.

**Reply by dhendry** [reply]
@bayandin The key point to me is the postgres version is already encoded higher up the path in multiple places so I would not think the additional `postgresql@14` is necessary:

`/opt/homebrew/Cellar/**postgresql@14**/**14.5_1**/lib/` -> I would never expect `postgresql@15` to be a valid subdirectory here

I am not sure how exactly `psycopg2` is getting the list of paths to search for but it looks like a pretty standard set. Moreover, the has recently broken with the latest postgresql@14 rename in homebrew so it seems to have changed in an incompatible way.

**Reply by carlocab** [reply]
Looks like this is expected behaviour then. Please use `pg_config` to find the right library directory. Alternatively, just use one of
```
$(brew --prefix postgresql@14)/lib/postgresql@14
```
or
```
$(brew --prefix)/lib/postgresql@14
```
or
```
$(brew --prefix)/opt/postgresql@14/lib/postgresql@14
```

**Reply by dhendry** [reply]
@bayandin also, there is not the additional level of version redirection in the `bin` directory:

```
➜  ~ ls -Alh /opt/homebrew/Cellar/postgresql@14/14.5_1/lib
total 0
drwxr-xr-x  106 dhendry  admin   3.3K 29 Aug 08:01 postgresql@14

➜  ~ ls -Alh /opt/homebrew/Cellar/postgresql@14/14.5_1/bin
total 25768
-r-xr-xr-x  1 dhendry  admin   113K 29 Aug 08:01 clusterdb
-r-xr-xr-x  1 dhendry  admin   113K 29 Aug 08:01 createdb
-r-xr-xr-x  1 dhendry  admin   113K 29 Aug 08:01 createuser
-r-xr-xr-x  1 dhendry  admin   112K 29 Aug 08:01 dropdb
-r-xr-xr-x  1 dhendry  admin   112K 29 Aug 08:01 dropuser
-r-xr-xr-x  1 dhendry  admin   873K  8 Aug 14:59 ecpg
-r-xr-xr-x  1 dhendry  admin   167K 29 Aug 08:01 initdb
...
```

**Reply by dhendry** [reply]
@carlocab @bayandin If I look at any other versioned package, there is not the same level of indirection. Consider openssl:

```
➜  ~ ls -Alh /opt/homebrew/Cellar/openssl@3/3.0.5/lib 
total 21824
drwxr-xr-x  11 dhendry  admin   352B  6 Jul 05:04 .
drwxr-xr-x  14 dhendry  admin   448B  6 Jul 05:04 ..
drwxr-xr-x   5 dhendry  admin   160B  6 Jul 05:04 engines-3
-rw-r--r--   1 dhendry  admin   3.2M  6 Jul 05:04 libcrypto.3.dylib
-r--r--r--   1 dhendry  admin   6.1M  5 Jul 02:57 libcrypto.a
lrwxr-xr-x   1 dhendry  admin    17B  5 Jul 02:57 libcrypto.dylib -> libcrypto.3.dylib
-r--r--r--   1 dhendry  admin   553K  6 Jul 05:04 libssl.3.dylib
-r--r--r--   1 dhendry  admin   861K  5 Jul 02:57 libssl.a
lrwxr-xr-x   1 dhendry  admin    14B  5 Jul 02:57 libssl.dylib -> libssl.3.dylib
drwxr-xr-x   3 dhendry  admin    96B  6 Jul 05:04 ossl-modules
drwxr-xr-x   5 dhendry  admin   160B  6 Jul 05:04 pkgconfig
```

The path is NOT `/opt/homebrew/Cellar/openssl@3/3.0.5/lib/openssl@3`

**Reply by carlocab** [reply]
The `Cellar` path is an implementation detail that you should not be referencing, because these paths are fragile and can change at any point. Please use `/opt/homebrew/lib/postgresql@14` instead.

The comparison to OpenSSL (or any other versioned formula) is not analogous because other formulae do not regularly ship major version bumps that break things for users.

**Reply by dhendry** [reply]
@carlocab @bayandin @MikeMcQuaid One last attempt to explain why I think there is a bug here:

1. I believe the reason psycopg2 is searching the `/opt/homebrew/Cellar/postgresql@14/14.5_1/lib/` path to begin with is that:
   1. It looks at the symlinked resolved path for the postgres executable
   2. It then uses the `lib` directory which is a sibling of the `bin` directory. 
   * This does not seem unreasonable to me based on my understanding of how unix directories should be laid out.

Here is a hacky bash equivalent of what I _believe_ the resolution logic is doing:

```bash
# Where is postgres:
➜  which postgres
/opt/homebrew/bin/postgres

# Whats the symlink its pointing at:
➜  ls -l  /opt/homebrew/bin/postgres
lrwxr-xr-x  1 dhendry  admin  43 29 Aug 08:01 /opt/homebrew/bin/postgres -> ../Cellar/postgresql@14/14.5_1/bin/postgres

# Resolve symlink: 
➜  realpath $(which postgres)
/opt/homebrew/Cellar/postgresql@14/14.5_1/bin/postgres

# lib directory path which is a sibling of the bin directory which seems reasonable:
➜  echo "$(realpath $(which postgres) | xargs dirname)/../lib/"
/opt/homebrew/Cellar/postgresql@14/14.5_1/bin/../lib/

# Contents of this lib directory (sibling of bin): 
# I would expect proper library contents to belong here but instead they are buried a directory deeper 
# (which seems unnecessary and incorrect, or at the very least very unexpected and non-standard)
➜  ls -l  /opt/homebrew/Cellar/postgresql@14/14.5_1/bin/../lib/
total 0
drwxr-xr-x  106 dhendry  admin  3392 29 Aug 08:01 postgresql@14
```

2. The logic I described above works for **every** versioned brew package I have checked so far including:
* `openssl@1.1`/`openssl@3`
* `llvm@11`
* `mysql`/`mysql@5.7`
* `mariadb`/`mariadb@10.7`

I have not seen a single case where any other package adds an additional level of version redirection under the already versioned cellar path

You can check for yourself with the following: `ls -ld "$(realpath $(which mysql) | xargs dirname)/../lib/"*`

For example:

```bash
# Check for mysql
➜  ls -ld "$(realpath $(which mysql) | xargs dirname)/../lib/"* | head   
-rw-r--r--    1 dhendry  admin  6822224 28 Jul 06:33 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlclient.21.dylib
-r--r--r--    1 dhendry  admin  7377720  6 Jul 16:01 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlclient.a
lrwxr-xr-x    1 dhendry  admin       23  6 Jul 16:01 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlclient.dylib -> libmysqlclient.21.dylib
-r--r--r--    1 dhendry  admin   861808 28 Jul 06:33 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlharness.1.dylib
-r--r--r--    1 dhendry  admin    57488 28 Jul 06:33 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlharness_stdx.1.dylib
-r--r--r--    1 dhendry  admin   101280 28 Jul 06:33 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlharness_tls.1.dylib
-r--r--r--    1 dhendry  admin  7896464 28 Jul 06:33 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlrouter.1.dylib
-r--r--r--    1 dhendry  admin   110992 28 Jul 06:33 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlrouter_connection_pool.1.dylib
-r--r--r--    1 dhendry  admin   129872 28 Jul 06:33 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlrouter_http.1.dylib
-r--r--r--    1 dhendry  admin   134032 28 Jul 06:33 /opt/homebrew/Cellar/mysql/8.0.30/bin/../lib/libmysqlrouter_http_auth_backend.1.dylib

# Same logic does NOT work for postgres (but it DOES for *every* other brew binary I have tried which has a lib companion):
➜  ls -ld "$(realpath $(which postgres) | xargs dirname)/../lib/"* | head
drwxr-xr-x  106 dhendry  admin  3392 29 Aug 08:01 /opt/homebrew/Cellar/postgresql@14/14.5_1/bin/../lib/postgresql@14
```

Sorry to keep bringing this up but this issue is a huge pain for me as we do not control the psycopg2 library and need to use an older version for legacy reasons (which we are trying to migrate away from as a separate initiative). I have resorted to asking my team to do really crazy things like `sudo ln -s /opt/homebrew/opt/postgresql@14/lib/postgresql@14/libpq.5.dylib /usr/local/lib`. 

Irrespective of my specific issue, this still just feels like a bug in the postgres homebrew formula. It seems entirely reasonable to me that the `lib` sibling of the `bin` directory has the same structure.

Thanks for the help and thanks for all your effort on the homebrew project!

**Reply by bayandin** [reply]
1. `psycopg2-binary` doesn't work because when they build binaries, `postgresql` formula libraries were in different directories. You can ask them to rebuild and publish a new version of the package, or you can use `psycopg2` (https://www.psycopg.org/docs/install.html#psycopg-vs-psycopg-binary).
2. Comparison with other formulae is irrelevant here. In https://github.com/Homebrew/homebrew-core/pull/107726 we explained why do we need a special behaviour for Postgres.

If you think this is a bug in the formula, we'll review the PR with a fix (if it won't break anything else).

**Reply by filbranden** [reply]
I also encountered this issue while using Ruby + the pg gem on MacOS on GitHub Actions, I reported it at actions/runner-images#6173.

The extra level `lib/postgresql@14` seems to be part of the problem, but not sure whether the top-level should have an unversioned symlink. (I haven't dug down into how that gem is finding the `/usr/local/opt/postgresql` path to begin with, so not sure about that part.)

**Reply by carlocab** [reply]
@filbranden try
```
LDFLAGS="-L$(pg_config --libdir)" CPPFLAGS="$(pg_config --cppflags)" gem install pg
```

**Reply by filbranden** [reply]
@carlocab It turns out the issue I was having was due to the bundle cache of the `ruby/setup-ruby` action. It had cached the gems built in the previous image, with the previous directory layout of the Postgres homebrew install, but then it was trying to use the same binaries on the new layout.

In debugging the issue, I noticed that building from source (even without explicit calls to pg_config, which the gem itself is already supposed to do anyways) would fix the problem correctly, so then I found more about the bundle cache and how to [force a rebuild](https://github.com/ruby/setup-ruby#dealing-with-a-corrupted-cache) which is needed here.

I solved that with https://github.com/filbranden/vimgolf/commit/fa2a5a80e4f6f88e6343c2caca2e7cc3815567b3 and running it through GitHub Actions confirms [it fixed the issue](https://github.com/filbranden/vimgolf/runs/8141356065?check_suite_focus=true).

Thanks for your help!

**Reply by filbranden** [reply]
I guess it's a little bit unfortunate that Homebrew is not keeping a stable ABI here, but I guess that's par for the course with Homebrew.

Also a little disappointing that this was very hard to troubleshoot, even though I'm pretty seasoned with build systems in general (so I guess TIL about this bundle cache, which is a nice one to know for the future!)

Anyways, I'm glad this is resolved now. Onwards.

**Reply by carlocab** [reply]
Well, the change was an attempt to make postgres installations more stable -- they tend to break when postgres changes its major version. While this much pain now was not anticipated, the idea was to cause less in the future (i.e. when postgres 15 comes along).

Glad you're sorted.

**Reply by dvarrazzo** [reply]
@carlocab Are you trying to solve problems on version changes on the server or on the client?

**Reply by carlocab** [reply]
We're trying to solve problems like this: #2285

**Reply by dvarrazzo** [reply]
That's a server problem, then. Moving the location of the libpq is something that affects the clients instead.

The libpq *might* be used by the server (for instance in a foreign data wrapper to another postgres instance, maybe?) but its main usage is for client programs to connect to a server. The libpq version doesn't matter: the higher the better, but current libpq can connect to any postgres server since 7.4.

Maybe you can install the libpq library as part of the [libpq brew package](https://formulae.brew.sh/formula/libpq) and then having the [postgres brew package](https://formulae.brew.sh/formula/postgresql@14) depending on it?

**Reply by carlocab** [reply]
> Maybe you can install the libpq library as part of the [libpq brew package](https://formulae.brew.sh/formula/libpq) and then having the [postgres brew package](https://formulae.brew.sh/formula/postgresql@14) depending on it?

That would be ideal, but I don't know of a way to configure `postgresql@14` to do that.

**Comment by ghost** [comment]
Going off previous suggestions to use `pg_config`, I put the following in my `~/.zprofile` (could also go in your `~/.zshrc`, or whatever your shell's configuration file is defined as):

```sh
export DYLD_FALLBACK_LIBRARY_PATH=`pg_config --libdir`
```

`DYLD_FALLBACK_LIBRARY_PATH` wasn't yet set to anything in my shell environment, but it's worth checking if it already is in yours.

**Reply by carlocab** [reply]
I don't recommend the use of `DYLD_FALLBACK_LIBRARY_PATH`, and neither does Apple. Borrowing from my comment at #3584:

`DYLD_FALLBACK_LIBRARY_PATH` may work for some things, but not others. Apple doesn't really like it either:
```
❯ FOO=foo /bin/bash -c 'echo ${FOO-empty}'
foo
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /bin/bash -c 'echo ${DYLD_FALLBACK_LIBRARY_PATH-empty}'
empty
❯ FOO=foo /usr/bin/python3 -c 'import os; print(os.getenv("FOO"))'
foo
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /usr/bin/python3 -c 'import os; print(os.getenv("DYLD_FALLBACK_LIBRARY_PATH"))'
None
```
The only way to avoid the behaviour above is to disable SIP, or to only use binaries that are not protected by SIP:
```
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /usr/local/bin/bash -c 'echo ${DYLD_FALLBACK_LIBRARY_PATH-empty}'
/usr/lib
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /usr/bin/env /usr/local/bin/bash -c 'echo ${DYLD_FALLBACK_LIBRARY_PATH-empty}'
empty
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /usr/local/bin/python3 -c 'import os; print(os.getenv("DYLD_FALLBACK_LIBRARY_PATH"))'
/usr/lib
```

**Reply by ghost** [reply]
Understood, I will resort to symlinking to `/usr/local/lib` for now (I'm on a M1 mac) to get `psycopg2` to work:

```sh
sudo mkdir -p /usr/local/lib && sudo ln -s /opt/homebrew/opt/postgresql@14/lib/postgresql@14/libpq.5.dylib /usr/local/lib/libpq.5.dylib
```

**Reply by dgramop** [reply]
I also had to symlink libpq.dylib (was getting Rust's diesel library to work, for context). Ran: 

```
sudo mkdir -p /usr/local/lib && sudo ln -s /opt/homebrew/opt/postgresql@14/lib/postgresql@14/libpq.5.dylib /usr/local/lib/libpq.5.dylib
 && sudo ln -s /opt/homebrew/opt/postgresql@14/lib/postgresql@14/libpq.dylib /usr/local/lib/libpq.dylib
```
